### PR TITLE
More documentation bits

### LINKED ISSUE #15497: Improve documentation of entity logic
http://tools.ietf.org/html/rfc7230#section-3.3.3 defines very clear rules for the entity length of HTTP messages.

Especially this rule requires special treatment in akka-http-core:

> Any response to a HEAD request and any response with a 1xx
> (Informational), 204 (No Content), or 304 (Not Modified) status
> code is always terminated by the first empty line after the
> header fields, regardless of the header fields present in the
> message, and thus cannot contain a message body.

Responses to HEAD requests introduce the complexity that `Content-Length` or `Transfer-Encoding` headers can be present but the entity is empty.
We model that by allowing the `HttpEntity.Default` and `HttpEntity.Chunked` to be used for HEAD responses with an empty data stream.

Also, when a HEAD response has an `HttpEntity.CloseDelimited` entity the `akka-http-core` layer will *not* close the connection after the response has been sent. This allows the sending of HEAD responses without `Content-Length` header across persistent HTTP connections.

These special cases need to be clearly described in the documentation.

### LINKED ISSUE #16812: Document json4s marshaller support
Make sure to provide a type-class based solution that prevents that the json4s marshallers will be used for any kind of data.

See spray/spray#990 for an issue of the old spray based design.

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #15497. Mostly copied from @sirthias description in the ticket.

Fixes #16812 and #17591.

### CODE REVIEW

File: akka-docs-dev/rst/java/http/client-side/host-level.rst [hunk 1]
```diff
@@ -80,6 +80,10 @@ response with its respective request. The way that this is done is by allowing t
 This context object of type ``T`` is completely opaque to Akka HTTP, i.e. you can pick whatever works best for your
 particular application scenario.
 
+.. note::
+  A consequence of using a pool is that long-running requests block a connection while running and may starve other
+  requests. Make sure not to use a connection pool for long-running requests like long-polling GET requests.
+  Use the :ref:`connection-level-api-java` instead.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/client-side/host-level.rst]: good note!

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3267/:

```
 10674 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 10675 Failed tests:
 10676 	akka.stream.scaladsl.GraphUnzipWithSpec
 14032 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 14033 Total time: 885 s, completed Jul 21, 2015 3:07:18 PM
```

**Comment by jrudolph** [comment]
(I'm testing a new simple error extraction script that quickly parses jenkins output for error lines.)

**Comment by ktoso** [comment]
> (I'm testing a new simple error extraction script that quickly parses jenkins output for error lines.)

That'd be awesome! Let me know how I can help or when it's ready so we can put it onto jenkins :)

**Comment by ktoso** [comment]
Nice additions, thanks!

**Pull request merged**